pkg/controller/constrainttemplate/constrainttemplate_controller_test.go

```diff
@@ -211,15 +222,48 @@ violation[{"msg": "denied!"}] {
 		Name:      "FooNamespace",
 		Object:    runtime.RawExtension{Object: ns},
 	}
-	resp, err := opa.Review(context.TODO(), req)
+	resp, err := opa.Review(ctx, req)
 	g.Expect(err).NotTo(gomega.HaveOccurred())
 	if len(resp.Results()) != 1 {
 		fmt.Println(resp.TraceDump())
-		fmt.Println(opa.Dump(context.TODO()))
+		fmt.Println(opa.Dump(ctx))
 	}
 	g.Expect(len(resp.Results())).Should(gomega.Equal(1))
 
+	// If the CRD is deleted out from underneath the template, it is recreated
+	log.Info("TESTING CRD RECREATION ON DELETE")
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+	crd.SetName(crdKey.Name)
+	crd.Spec = apiextensionsv1beta1.CustomResourceDefinitionSpec{}
+	g.Expect(c.Delete(ctx, crd)).NotTo(gomega.HaveOccurred())
+
+	g.Eventually(func() error {
+		// Drain the request queue so the controller can operate
+		select {
+		case v := <-requests:
+			log.Info("Reconciling request", "request", v)
+		default:
+		}
+		crd := &apiextensionsv1beta1.CustomResourceDefinition{}
+		if err := c.Get(ctx, crdKey, crd); err != nil {
+			return err
+		}
+		if !crd.GetDeletionTimestamp().IsZero() {
+			return errors.New("Still deleting")
+		}
+		for _, cond := range crd.Status.Conditions {
+			if cond.Type == apiextensionsv1beta1.Established && cond.Status == apiextensionsv1beta1.ConditionTrue {
+				return nil
+			}
+		}
+		return errors.New("Not established")
+	}, timeout, time.Second).Should(gomega.BeNil())
+
+	g.Eventually(func() error { return c.Create(ctx, newDenyAllCstr()) }, timeout).Should(gomega.BeNil())
+	constraintEnforced(g)
+
 	// Create template with invalid rego, should populate parse error in status
+	log.Info("TESTING INVALID REGO")
```

Comment:
I think it's worth refactoring some of these as separate / subtests. I can create an issue to track that.

Reply:
+1 definitely getting ugly.

Reply:
refactored